cxx/gendb.cc

```diff
@@ -160,24 +165,59 @@ void GenDB::incorporate_query_relation(std::mt19937* prng,
         },
         query_rel);
 
-    bool base_contains_items = std::visit(
-        [&](auto rel) { return rel->get_data().contains(base_items); },
-        hirm->get_relation(t_query_rel->base_relation));
-    if (!base_contains_items) {
-      hirm->sample_and_incorporate_relation(prng, t_query_rel->base_relation,
-                                            base_items);
-    }
+    const std::vector<std::string>& domains =
+        std::visit([&](auto tr) { return tr.domains; },
+                   hirm->schema.at(t_query_rel->base_relation));
+    sample_and_incorporate_for_class(prng, domains.back(), base_items.back());
   }
   hirm->incorporate(prng, query_rel_name, items, value);
 }
 
+// Samples and incorporates a value into all relations belonging to class_name
+// (including class attributes and noisy observations of ancestor class
+// attributes). Recursively calls itself on base relations.
+void GenDB::sample_and_incorporate_for_class(std::mt19937* prng,
```

Comment:
Just curious -- will this work when class_name is the name of the observation class?  Because if so, we could use it to slightly simplify the implementation of make_pclean_sample in pclean_lib.

Reply:
Good point, this should work for the observation class too.